### PR TITLE
Router - document published metrics

### PR DESCRIPTION
**Pull request by pecameron**
OCP 3.6

Documents metrics that are published by the router
Document installation and configuration
Document related environment variables

Trello:
https://trello.com/c/EHq8dqNj/529-document-the-metrics-the-router-exposes

### CODE REVIEW

**Comment by pecameron** [comment]
@knobunc @mburke5678 @bfallonf PTAL

**Comment by pecameron** [comment]
Just squashed the early commit.

File: architecture/topics/haproxy_template_router.adoc [hunk 1]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
+The HAProxy router, by default, exposes or publishes metrics in
+link:https://prometheus.io/docs/concepts/data_model/[Prometheus format]
+for consumption by external metrics collection and aggregation systems (e.g. Prometheus, statsd).
+The router can be
+xref:../../install_config/router/default_haproxy_router.adoc#exposing-the-router-metrics[confiugred]
+to provide
+link:https://cbonte.github.io/haproxy-dconv/1.5/configuration.html#9[HAProxy CSV format] metrics, or
+provide no router metrics at all.
+
+Prometheus format data is the default data format for metrics in {product-title}.
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: I'm not sure we need to say this.

Or can we integrate it with the previous paragraph.

File: architecture/topics/haproxy_template_router.adoc [hunk 2]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
 
 .HAProxy Router Data Flow
-image::router_model.png[HAProxy Router Data Flow]
\ No newline at end of file
+image::router_model.png[HAProxy Router Data Flow]
+
+[[haproxy-metrics]]
+=== HAProxy Template Router Metrics
+
+The HAProxy router, by default, exposes or publishes metrics in
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: "by default"

Not sure if we want to say that... why not drop it.

File: architecture/topics/haproxy_template_router.adoc [hunk 3]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
+for consumption by external metrics collection and aggregation systems (e.g. Prometheus, statsd).
+The router can be
+xref:../../install_config/router/default_haproxy_router.adoc#exposing-the-router-metrics[confiugred]
+to provide
+link:https://cbonte.github.io/haproxy-dconv/1.5/configuration.html#9[HAProxy CSV format] metrics, or
+provide no router metrics at all.
+
+Prometheus format data is the default data format for metrics in {product-title}.
+
+The metrics are scraped from the the router controller and from HAProxy.
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: The metrics are from both the router controller, and contain the statistics retrieved from the HAProxy that the router controls.

File: architecture/topics/haproxy_template_router.adoc [hunk 4]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
+
+The router scrapes the following HAProxy metrics every 5 seconds. To reduce resource usage when there are
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: s/scrapes/gathers/

File: architecture/topics/haproxy_template_router.adoc [hunk 5]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
+more than 500 servers, the backends are reported instead of the servers.
+
+The router metrics start at zero when the router is deployed. The HAProxy metrics are reset to
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: s/metrics/counters/?

File: architecture/topics/haproxy_template_router.adoc (hunk 4 above)
Comment (knobunc) [reply architecture/topics/haproxy_template_router.adoc]: This whole paragraph feels like it should be combined with the one that is two below.

File: architecture/topics/haproxy_template_router.adoc (hunk 3 above)
Comment (knobunc) [reply architecture/topics/haproxy_template_router.adoc]: This whole paragraph feels like it should be combined with the one that is two below.

File: architecture/topics/haproxy_template_router.adoc [hunk 6]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
+zero every time haproxy is reloaded.
+
+The router scrapes HAProxy statistics for each frontend, backend and server. The statistics are
+a subset of the available HAProxy
+link:https://cbonte.github.io/haproxy-dconv/1.5/configuration.html#9.1[Statistics].
+
+The following HAProxy metrics are scraped on a preiodic basis and converted to Prometheus format.
+For every frontend the "F" counters are scraped. When the total number of servers is less than
+the Server Threshold, the "b" counters are scraped for each backend and the "S" server counters
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: Server Threshold (default 500)

File: architecture/topics/haproxy_template_router.adoc [hunk 7]
```diff
@@ -27,4 +27,106 @@ The following diagram illustrates how data flows from the master through the
...
+
+Column 2
+|===
+|F|Frontend metrics
+|b|backend metrics when not showing Server metrics due to the Server Threshold,
+|B|backend metrics when showing Server metrics
+|S|Server metrics.
+|===
+
+Column 3 - the counter
```
Comment (knobunc) [inline architecture/topics/haproxy_template_router.adoc]: Inconsistent capitalization across these descriptions

**Comment by mburke5678** [comment]
@pecameron Overall, this looks good. A nice addition. 

We have some information on HAProxy router and template routers that appears to have been added in 3.6. We could probably combine with your doc.
https://docs.openshift.com/container-platform/3.6/architecture/networking/routes.html#routes-template-routers

I came across this blog from @cpitman about creating a GUI for the HAProxy router metrics (which I haven't gotten to work yet; but I am sure that it is just me). Is this applicable for 3.6 and something we want to advertise in the official docs?
https://docs.openshift.com/container-platform/3.6/architecture/networking/routes.html#routes-template-routers

File: architecture/topics/haproxy_template_router.adoc (hunk 2 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: Changed

File: architecture/topics/haproxy_template_router.adoc (hunk 1 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: Dropped line

File: architecture/topics/haproxy_template_router.adoc (hunk 3 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: reworded

File: architecture/topics/haproxy_template_router.adoc (hunk 4 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: reorganized discussion

File: architecture/topics/haproxy_template_router.adoc (hunk 5 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: OK

File: architecture/topics/haproxy_template_router.adoc (hunk 6 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: OK

File: architecture/topics/haproxy_template_router.adoc (hunk 7 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: OK

**Comment by pecameron** [comment]
@gaurav-nelson @mburke5678 @knobunc PTAL

File: install_config/router/default_haproxy_router.adoc [hunk 8]
```diff
@@ -1500,109 +1500,99 @@ add routes across the namespaces.
...
 ----
-$ oc adm router --service-account=router --expose-metrics
+$ oc adm router --service-account=router
 ----
 ====
-+
-Or, optionally, use the `--metrics-image` option to override the HAProxy
-defaults:
-+
+metrics are automatically available in prometheus format on the stats-port, default 1936. To suppress metrics collection,
```
Comment (mburke5678) [inline install_config/router/default_haproxy_router.adoc]: s/prometheus/Prometheus

File: install_config/router/default_haproxy_router.adoc [hunk 9]
```diff
@@ -1500,109 +1500,99 @@ add routes across the namespaces.
...
-. Create the {product-title} router:
-+
+[NOTE]
 ====
-----
-$ oc adm router --service-account=router --expose-metrics
-----
+The `--expose-metrics` and `--metrics-image` options are deprecated. The haproxy-exporter
+side car is now integrated into the router controller so you can delete the sidecar from existing
+router DCs. You can continue to use the side car in existing routers. New routers use the integrated metrics.
```
Comment (mburke5678) [inline install_config/router/default_haproxy_router.adoc]: Elsewhere we refer to `sidecar container` and not just `sidecar`. Is this the same concept? Maybe use `sidecar container` here?
https://docs.openshift.org/latest/install_config/router/default_haproxy_router.html#exposing-the-router-metrics

File: install_config/router/default_haproxy_router.adoc [hunk 10]
```diff
@@ -1500,109 +1500,99 @@ add routes across the namespaces.
...
 ====
-+
-Or, optionally, use the `--metrics-image` option to override the HAProxy
-defaults:
-+
+
+
+You can extract the raw statistics in prometheus format by using the following.
```
Comment (mburke5678) [inline install_config/router/default_haproxy_router.adoc]: s/prometheis/Prometheus

File: architecture/topics/router_environment_variables.adoc [hunk 11]
```diff
@@ -44,8 +44,10 @@ connections (and any time HAProxy is reloaded), the old HAProxy processes
 will "linger" around for that period. xref:time-units[(TimeUnits)]
 |`ROUTER_DENIED_DOMAINS` | | A comma-separated list of domains that the host name in a route can not be part of. No subdomain in the domain can be used either. Overrides option `ROUTER_ALLOWED_DOMAINS`.
 |`ROUTER_ENABLE_COMPRESSION`| | If `true` or `TRUE`, compress responses when possible.
+|`ROUTER_LISTEN_ADDR`| 0.0.0.0:1936 | Sets the listening address for xref:../../install_config/router/default_haproxy_router.adoc#exposing-the-router-metrics[router metrics].
 |`ROUTER_LOG_LEVEL` | warning | The log level to send to the syslog server.
 |`ROUTER_MAX_CONNECTIONS`| 20000 | Maximum number of concurrent connections.
+|`ROUTER_METRICS_TYPE`| haproxy | Generate metrics for xref:../../install_config/router/default_haproxy_router.adoc#exposing-the-router-metrics[the haproxy router]. (haproxy is the only supported value)
 |`ROUTER_OVERRIDE_HOSTNAME`|  | If set `true`, override the spec.host value for a route with the template in `ROUTER_SUBDOMAIN`.
```
Comment (mburke5678) [inline architecture/topics/router_environment_variables.adoc]: s/haproxy/`haproxy`

**Comment by mburke5678** [comment]
@pecameron A couple of comments; I can do a follow-up PR for the edits, if you prefer.

File: install_config/router/default_haproxy_router.adoc (hunk 8 above)
Comment (pecameron) [reply install_config/router/default_haproxy_router.adoc]: Changed

File: install_config/router/default_haproxy_router.adoc (hunk 9 above)
Comment (pecameron) [reply install_config/router/default_haproxy_router.adoc]: Changed

File: install_config/router/default_haproxy_router.adoc (hunk 10 above)
Comment (pecameron) [reply install_config/router/default_haproxy_router.adoc]: Changed

File: architecture/topics/router_environment_variables.adoc (hunk 11 above)
Comment (pecameron) [reply architecture/topics/router_environment_variables.adoc]: Not sure what you want to change here...

**Comment by pecameron** [comment]
@mburke5678 PTAL

File: architecture/topics/router_environment_variables.adoc (hunk 11 above)
Comment (mburke5678) [reply architecture/topics/router_environment_variables.adoc]: @pecameron Sorry. Maybe put the ` characters around haproxy.

File: architecture/topics/haproxy_template_router.adoc [hunk 12]
```diff
@@ -27,4 +27,102 @@ The following diagram illustrates how data flows from the master through the
...
+|template_router_write_config_seconds|Measures the time spent writing out the router configuration to disk in seconds.
+|haproxy_exporter_up|Was the last scrape of haproxy successful.
+|haproxy_exporter_csv_parse_failures|Number of errors while parsing CSV.
+|haproxy_exporter_scrape_interval|The time in seconds before another scrape is allowed, proportional to size of data.
+|haproxy_exporter_server_threshold|Number of servers tracked and the current threshold value.
+|haproxy_exporter_total_scrapes|Current total HAProxy scrapes.
+|http_request_duration_microseconds|The HTTP request latencies in microseconds.
+|http_request_size_bytes|The HTTP request sizes in bytes.
+|http_response_size_bytes|The HTTP response sizes in bytes.
+|openshift_build_info|A metric with a constant '1' value labeled by major, minor, git commit & git version from which OpenShift was built.
```
Comment (mburke5678) [inline architecture/topics/haproxy_template_router.adoc]: Not sure what "from which OpenShift was built." Which cluster??
If using OpenShift, please use {product-title}

File: architecture/topics/haproxy_template_router.adoc [hunk 13]
```diff
@@ -27,4 +27,102 @@ The following diagram illustrates how data flows from the master through the
...
+|ssh_tunnel_open_count|Counter of ssh tunnel total open attempts
+|ssh_tunnel_open_fail_count|Counter of ssh tunnel failed open attempts
```
Comment (mburke5678) [inline architecture/topics/haproxy_template_router.adoc]: s/ssh tunnel/SSH tunnel

File: install_config/router/default_haproxy_router.adoc [hunk 14]
```diff
@@ -1500,109 +1500,99 @@ add routes across the namespaces.
...
-----
-$ sudo docker pull prom/haproxy-exporter
-----
-====
+The
+xref:../../architecture/networking/haproxy-router.adoc#haproxy-metrics[HAProxy router metrics]
+are, by default, exposed or published in
+link:https://prometheus.io/docs/concepts/data_model/[Prometheus format]
+for consumption by external metrics collection and aggregation systems (e.g. Prometheus, statsd)..
+Metrics are also available dirctly from the
```
Comment (mburke5678) [inline install_config/router/default_haproxy_router.adoc]: (e.g. Prometheus, statsd)..
Two periods

File: install_config/router/default_haproxy_router.adoc [hunk 15]
```diff
@@ -1500,109 +1500,99 @@ add routes across the namespaces.
...
-. Create the {product-title} router:
-+
+[NOTE]
 ====
-----
-$ oc adm router --service-account=router --expose-metrics
-----
+The `--expose-metrics` and `--metrics-image` options are deprecated. The haproxy-exporter
+side car is now integrated into the router controller so you can delete the sidecar container from existing
+router DCs. You can continue to use the side car in existing routers. New routers use the integrated metrics.
```
Comment (mburke5678) [inline install_config/router/default_haproxy_router.adoc]: s/router DCs/router deployment configs

**Comment by mburke5678** [comment]
@pecameron Text looks good. Did you work with @bfallonf to make sure you edited the right topics files? I am not familiar with how they are mapped into other topics. Or, if they still are.

File: architecture/topics/haproxy_template_router.adoc (hunk 12 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: The git source control software has a version too. This just shows it.

File: architecture/topics/haproxy_template_router.adoc (hunk 13 above)
Comment (pecameron) [reply architecture/topics/haproxy_template_router.adoc]: Changed

File: install_config/router/default_haproxy_router.adoc (hunk 14 above)
Comment (pecameron) [reply install_config/router/default_haproxy_router.adoc]: changed

File: install_config/router/default_haproxy_router.adoc (hunk 15 above)
Comment (pecameron) [reply install_config/router/default_haproxy_router.adoc]: Changed

**Comment by pecameron** [comment]
@mburke5678 I made the suggested changes. PTAL

**Comment by pecameron** [comment]
@mburke5678  has this information been merged in a different PR?

**Comment by pecameron** [comment]
@mburke5678, @mrobson has been looking for the information in our docs.

**Comment by mburke5678** [comment]
@pecameron I am not sure if you made the changes in the right .adoc files. 

@bfallonf Can you look at what Phil did? I think the files he used are yours??

**Comment by mburke5678** [comment]
@pecameron 
@bfallonf says we are good to merge this. What version(s) of OpenShift does this apply to?

**Comment by pecameron** [comment]
@mburke5678 @bfallonf  - 3.6 and newer

**Comment by mburke5678** [comment]
[rev_history]
|xref:../architecture/networking/haproxy-router.adoc#architecture-core-concepts-haproxy-router[Architecture -> HAProxy Router Plug-in]
|New section describing the xref:../architecture/networking/haproxy-router.adoc#haproxy-template-router-metrics[HAProxy Template Router Metrics].
%
|xref:../architecture/networking/routes.adoc#architecture-core-concepts-routes[Architecture -> Routes]
|Added information on the `ROUTER_LISTEN_ADDR` and `ROUTER_METRICS_TYPE` variables to the xref:../architecture/networking/routes.adoc#router-environment-variables[Router Environment Variables] section.
%
|xref:../install_config/router/default_haproxy_router.adoc#install-config-router-default-haproxy[Using the Default HAProxy Router]
|Modified some information in the xref:../install_config/router/default_haproxy_router.adoc#exposing-the-router-metrics[Router Environment Variables] section.
%

**Pull request merged**